src/main/java/org/spongepowered/api/text/template/TextTemplate.java

```diff
@@ -0,0 +1,214 @@
+     * Evaluates this TextTemplate and returns the result in a
+     * {@link Text.Builder}.
+     *
+     * @param params Parameters to evaluate
+     * @return Text builder containing result
+     * @throws RequiredTextTemplateArgumentException if required parameters are missing
+     */
+    public Text.Builder eval(Map<String, Object> params) {
+        Text.Builder result = Text.builder();
+        for (Object element : this.elements) {
+            appendElement(result, element, params);
+        }
+        return result;
+    }
+
+    private void appendElement(Text.Builder builder, Object element, Map<String, Object> params) {
+        if (element instanceof TextElement) {
+            ((TextElement) element).appendTo(builder);
+        } else if (element instanceof Arg) {
+            Arg<?> arg = (Arg<?>) element;
+            Object param = params.get(arg.name);
+            if (param == null) {
+                if (!arg.optional) {
+                    throw new RequiredTextTemplateArgumentException(arg.name);
+                }
+            } else {
+                arg.eval(param).appendTo(builder);
+            }
+        } else {
+            builder.append(Text.of(element.toString()));
+        }
+    }
+
+    /**
+     * Sends the result of this TextTemplate with the specified parameters to
+     * the specified {@link MessageReceiver}.
+     *
+     * @param receiver Receiver to send message to
+     * @param params Parameters to apply to template
+     */
+    public void send(MessageReceiver receiver, Map<String, Object> params) {
+        receiver.sendMessage(eval(params).build());
+    }
+
+    /**
+     * Constructs a new TextTemplate for the given elements. The order of the
+     * elements is the order in which they will be appended to the result
+     * builder via {@link #eval(Map)}.
+     *
+     * The provided elements may be of any type.
+     *
+     * In the case that an element is a {@link TextElement},
+     * {@link TextElement#appendTo(Text.Builder)} will be used to append the
+     * element to the builder.
+     *
+     * In the case that an element is an {@link Arg} the argument will be
+     * replaced with a {@link TextElement} object via the {@link Function}
+     * provided by {@link #arg(String, Class, Function, boolean)}. If no
+     * function was provided the argument expects a TextElement type. The
+     * parameter value in the provided map in {@link #eval(Map)} must match the
+     * type of the input for the arguments Function or a
+     * {@link MismatchedTextTemplateArgumentException} will be thrown.
+     *
+     * In the case that an element is any other type, the parameter value's
+     * {@link Object#toString()} method will be used to create a {@link Text}
+     * object.
+     *
+     * @param elements Elements to append to builder
+     * @return Newly constructed TextTemplate
+     */
+    public static TextTemplate of(Object... elements) {
+        return new TextTemplate(elements);
```

Comment:
It would be good to have this do something special if elements is empty (e.g. return constant which results in an empty builder everytime or throw an exception)